python_ga/mutation.py

```diff
@@ -0,0 +1,130 @@
+                    break
+                break
+        population[i] = refine_genome_around_cross_point_to_time_constraint(
+            genome=genome, interval_in_indices=interval_in_indices)
+    return population
+
+
+def mutate_time_value(population, config):
+    """
+    Mutacja zmian losowych czasu podania dawki
+    :param population:
+    :param config:
+    :return:
+    """
+    interval_in_indices = int(2 * config['time_interval_hours'])
+
+    mutation_config = config['mutations']['mutate_time_value']
+    max_dose_value = config['max_dose_value']
+
+    for i, genome in enumerate(population):
+        for gene_idx in range(len(genome)):
+            p = np.random.uniform()
+            if p < mutation_config['mut_prob']:
+                if genome[gene_idx] > 0:
+                    new_dose_time = np.random.randint(len(genome))
+                    new_dose_value = genome[new_dose_time] + genome[gene_idx]
+                    dose_remainder = 0
+                    if new_dose_value > max_dose_value:
+                        dose_remainder = new_dose_value - max_dose_value
+                        new_dose_value = max_dose_value
+
+                    genome[new_dose_time] = new_dose_value
+                    genome[gene_idx] = dose_remainder
+        population[i] = refine_genome_around_cross_point_to_time_constraint(
+            genome=genome, interval_in_indices=interval_in_indices)
+    return population
+
+
+def mutate_swap(population, config):
+    """
+    Mutacja wymiany: badany jest każdy gen w genomie wszystkich osobników w populacji. Jeżeli wylosowana wartość p jest
+    większa niż określone prawdopodobieństwo mutacji, wybrany gen jest zamieniamy miejscami z losowo wybranym innym genem
+    :param population:      list
+    :param config:          dict
+    :return: population:    list
+    """
+    interval_in_indices = int(2 * config['time_interval_hours'])
+
+    mutation_config = config['mutations']['mutate_swap']
+    for i, genome in enumerate(population):
+        for gene_idx in range(len(genome)):
+            p = np.random.uniform()
+            if p < mutation_config['mut_prob']:
+                k = np.random.randint(len(genome) - 1)
+                k = (gene_idx + k) % len(genome)
+                genome[gene_idx], genome[k] = genome[k], genome[gene_idx]
+        population[i] = refine_genome_around_cross_point_to_time_constraint(
+            genome=population[i], interval_in_indices=interval_in_indices)
+    return population
+
+
+def mutate_split(population, config, max_dose=10, min_dose=0.25):
+    """
+    Splitting a non-zero dose (> 0.25Gy) into 2 doses.
+    population - next population, array [population_size, element_size].
+    """
+    interval_in_indices = int(2 * config['time_interval_hours'])
+
+    mutation_config = config['mutations']['mutate_merge']
+
+    population = np.asarray(population)
+    for i, genome in enumerate(population):
```

Comment:
lines 262-284 from source are rewritten to 107-127 here, to the same iteration as in `mutate_time_value` and `mutate_dose_value`

algorithm logic not changed